src/gui/editors/PianoRoll.cpp

```diff
@@ -3413,43 +3487,56 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
 		};
 
 		// -- Begin ghost MIDI clip
-		if( !m_ghostNotes.empty() )
+		if (!m_ghostNotes.empty() && m_ghostVisible.value())
 		{
-			for( const Note *note : m_ghostNotes )
+			// Lambda function that takes a value x and rounds it to the next/previous value evenly divisable by q
+			// a.k.a. quantizing... (unlike integer division this is the same for negative numbers)
+			auto roundDown = [](int x, int q) -> int { return std::floor(static_cast<float>(x) / q) * q; };
+			auto roundUp = [](int x, int q) -> int { return std::ceil(static_cast<float>(x) / q) * q; };
+
+			// Horizontal repeat
+			const bool loop = m_ghostRepeated.value();
+			// Round to whole bars
+			const int ghostLength = roundUp(m_ghostBounds.end - m_ghostBounds.start, TimePos::ticksPerBar());
+			// Offset of the leftmost visible repetition, relative to the original ghost notes
+			const int posOffBegin = loop ? roundDown(m_currentPosition - m_ghostBounds.start, ghostLength) : 0;
+			// Tick to the right of the visible area
+			const int rightmostTick = m_currentPosition + (width() - m_whiteKeyWidth) * TimePos::ticksPerBar() / m_ppb;
+			// Where to stop drawing
+			const int posEnd = loop ? rightmostTick : m_ghostBounds.end.getTicks();
+
+			// Vertical repeat
+			const bool stack = m_ghostStacked.value();
+			// Round to whole octaves
+			const int ghostRange = roundUp((m_ghostBounds.highest - m_ghostBounds.lowest + 1), KeysPerOctave);
```

Comment:
I'm not sure how vertical repeat in this way will be useful. Can the ghost notes be transposed in some way instead of this?

Reply:
Personally I'd use it in the same way you use the key/scale marker but with ghost notes you get more flexibility and can have a chord progression laid out.

Plus it's nice how you automatically have the ghost there no matter if you write bass or high.

If your ghost has a wide range I suppose it can end up on the wrong octave even if repeat is on. In that case the temporary solution would be paste ghost/move/apply as ghost. But we could add transpose buttons in another PR, it's a good idea.